### ISSUE TITLE
[Bug] Microsoft.Identity.Client 4.52.0 introduces dependency on Microsoft.NETCore.App AND Microsoft.WindowsDesktop.App

### ISSUE THREAD

**Issue by blaise-braye** [AzureAD/microsoft-authentication-library-for-dotnet#4164]
**Which version of MSAL.NET are you using?**
upgrading from version Microsoft.Identity.Client.Extensions.Msal version 2.25.3 to 2.31.0

**Platform**
<TargetFramework>net6.0-windows</TargetFramework>
<RuntimeIdentifier>win10-x64</RuntimeIdentifier>

**What authentication flow has the issue?**
None

Other?
The upgrade forces the netcore application to rely on the windows desktop framework where it used to not have that dependency.
It is problematic because the size of the base image is much bigger if we go for desktop images and, furthermore, it is not the intent...
by applying a bisect on the different dependencies in the upgrade path, we identified that the issue happens because of the package Microsoft.Identity.Client 4.52.0.

**Is this a new or existing app?**
yes, we could work around the problem by specicying manually the runtimeconfig.json file
however this is not a sustainable solution on the long run, it hides other potential compatibility conflicts.

**Repro**
```
$ dotnet new console -n repro -f net6.0-windows
$ (Get-Content .\repro\repro.csproj).Replace("<TargetFramework>net6.0</TargetFramework>", "<TargetFramework>net6.0-windows</TargetFramework><RuntimeIdentifier>win10-x64</RuntimeIdentifier><SelfContained>false</SelfContained>") | Set-Content .\repro\repro.csproj
$ dotnet build repro
$ cat .\repro\bin\Debug\net6.0-windows\win10-x64\repro.runtimeconfig.json
{
  "runtimeOptions": {
    "tfm": "net6.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "6.0.0"
    }
  }
}

$ dotnet add repro package Microsoft.Identity.Client --version 4.51.0
$ dotnet build repro
$ cat .\repro\bin\Debug\net6.0-windows\win10-x64\repro.runtimeconfig.json
{
    "runtimeOptions": {
        "tfm": "net6.0",
        "framework": {
            "name": "Microsoft.NETCore.App",
            "version": "6.0.0"
        }
    }
}
$ dotnet add repro package Microsoft.Identity.Client --version 4.52.0 # and above
$ dotnet build repro
$ cat .\repro\bin\Debug\net6.0-windows\win10-x64\repro.runtimeconfig.json
cat .\repro\bin\Debug\net6.0-windows\win10-x64\repro.runtimeconfig.json
{
    "runtimeOptions": {
        "tfm": "net6.0",
        "frameworks": [
            {
                "name": "Microsoft.NETCore.App",
                "version": "6.0.0"
            },
            {
                "name": "Microsoft.WindowsDesktop.App",
                "version": "6.0.0"
            }
        ]
    }
}
```

**Expected behavior**
runtimeconfig.json:
```
{
  "runtimeOptions": {
    "tfm": "net6.0",
    "frameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "6.0.0"
      }
    ]
  }
}
```

**Actual behavior**
runtimeconfig.json:
```
{
  "runtimeOptions": {
    "tfm": "net6.0",
    "frameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "6.0.0"
      },
      {
        "name": "Microsoft.WindowsDesktop.App",
        "version": "6.0.0"
      }
    ]
  }
}
```

**Comment by pmaytak** [comment]
In 4.52.0, we changed the `net6.0-windows.10.0.17763` target to `net6.0-windows`. We updated the Windows broker implementation which no longer required us to target the specific Windows SDK version. `net6.0` is available in pre- and post-4.52.0 MSAL.

So, if the client app targets `net6.0-windows`, it would use `net6.0` binaries for MSAL 4.52.0-, and `net6.0-windows` binaries for MSAL 4.53.0+.

Both `net6.0-windows.10.0.17763` and `net6.0-windows`, do reference Windows SDK. This is needed for MSAL to use the embedded browser for interactive authentication.

The guidance:
- For confidential client apps, or if you don't need interactive authentication, target `net6.0`.
- If you want to use embedded browser for authentication, target `net6.0-windows`. (Previously, you would have to target `net6.0-windows.10.0.17763`.)
- If you want to use [Windows broker](https://aka.ms/msal-net-wam), you can target either 'net6.0-windows` or `net6.0` and add a reference to Microsoft.Identity.Client.Broker.

Seems like you were ok using the `net6.0` MSAL. Does your app need to target `net6.0-windows`? If so, you may need to explicitly target `net6.0` MSAL in your scenario.

**Comment by blaise-braye** [comment]
Hello Peter, 
thank you for making the time to reply, very appreciated.

my application indeed needs to target net6.0-windows, main reason is that it executes user impersonation with the WindowsIdentity class coming from System.Security.Principal.Windows. This component won't let me target net6.0, we must be explicit about the target platform otherwise we get compilation errors.

I am not sure to understand your final remark, is it possible to force nuget to pick a specific target framework explicitly? do you refer to this approach maybe? https://duanenewman.net/blog/post/a-better-way-to-override-references-with-packagereference/
we could do that, it really feels like a workaround too at this stage... 
at least, it would make the deterministic runtime selection under control again, which is a decent move forward. we'll give it a try now.

wondering, wouldn't it make sense to fork the interaction components of msal in their own package, it seems that the package makes a strong decision around use case based on the target framework. 
as I showed you here, I have a use case here that is "fighting" against this design choice:
- no user interaction is expected
- the component can run on windows only
- the component does not need the full .net windows runtime to run

**Comment by blaise-braye** [comment]
Hello Peter,
coming with the outcome of the explicit selection of the net6 dependency:
```
<ItemGroup>
    <PackageReference Include="Microsoft.Identity.Client" Version="4.54.1" ExcludeAssets="Compile" GeneratePathProperty="true" />
  </ItemGroup>
  <ItemGroup>
    <Reference Include="Microsoft.Identity.Client">
      <HintPath>$(PkgMicrosoft_Identity_Client)\lib\net6.0\Microsoft.Identity.Client.dll</HintPath>
    </Reference>
  </ItemGroup>
```
unfortunately, the build process still implies unwanted dependency on the windows desktop runtime, ayn other idea is welcome.

{
  "runtimeOptions": {
    "tfm": "net6.0",
    "frameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "6.0.0"
      },
      {
        "name": "Microsoft.WindowsDesktop.App",
        "version": "6.0.0"
      }
    ]
  }
}

**Comment by pmaytak** [comment]
I was able to repro. Looks like the correct `net6.0` binary is used; but `net6.0-windows` is still used to restore the dependencies. I'm trying to get some clarifications about this.

> wouldn't it make sense to fork the interaction components of msal in their own package

We've explored this but the possible solutions end up bringing more complexity into the code base and packaging.

**Comment by bgavrilMS** [comment]
We will reach out to .NET folks for a comment, but we do not have a solution for this now.

**Comment by blaise-braye** [comment]
Hello,
ok great, please keep me in the loop, I would be interested to understand what happens under the hood.
We highly value having an up to date code base, mostly, to prevent any unwanted open security hole --> For the time being, we'll keep the dependency up to date and overwrite the runtimeconfig

Regards,
Blaise

**Comment by blaise-braye** [comment]
Hello, while in the progress of the migration of one of our components, an engineer found a way to workaround this better.
Around the question, can we target net6.0 instead of net6.0-windows: the answer becomes yes.

adding some context for anyone who would come in the same trap:
1. we were initally using the attribute [System.Runtime.Versioning.SupportedOSPlatform("windows")] to meet the code analyser  requirements. Indeed, WindowsIdentity class is not available elsewhere than on Windows.
2. by doing that, the build process would force us have a target framework defined as net6-windows --> that's where we ended.

The new information is that you can obtain a happy analyzer with a different trick: instead of using this attribute, you can check the operating system in your code like this:
```
if (OperatingSystem.IsWindows())
{
    return new WindowsIdentity(logonToken);
}
```

Roslyn is smart enough to observe that your code takes care of the operating system now. 
By design, it assumes then that your code becomes portable (which makes sense, when we think about it twice, from this angle, it was not the case before and it became the case by branching the code according to the operating system).
With this approach, we could target the framework net6.0 and the entire story is solved. 
This is a perfect solution for us, we don't need the desktop runtime anymore, because the code is actually written in a portable way now.

With the hope it can help others,

Best regards, 
Blaise

**Comment by bgavrilMS** [comment]
Thanks for posting the updated workaround @blaise-braye !

Will be closing this issue, as "Won't Fix"

**Comment by pmaytak** [comment]
We've released MSAL [4.61.0](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/4.61.0) which removes net6.0-windows binary (Windows Forms dependency).
For confidential client apps targeting net6.0-windows, there are no migration steps. You should be able to upgrade MSAL and remove the workarounds. The app will pick up net6.0 MSAL binary.
For public client apps targeting net6.0-windows, reference Microsoft.Identity.Client.Broker when using [interactive authentication with Windows Broker](https://aka.ms/msal-net-wam) and call WithBroker(BrokerOptions); or reference Microsoft.Identity.Client.Desktop when [authenticating with browser](https://aka.ms/msal-net-uses-web-browser) and call WithWindowsEmbeddedBrowserSupport(). There are no changes to the usage of the system browser.
